### PR TITLE
Add support for fast interval finder.

### PR DESCRIPTION
**Pull request by niyarin**
This PR adds support for the data structure and  function  for intervals that overlap the specified range (chr,start,end).

### CODE REVIEW

**Comment by alumi** [comment]
[image](https://github.com/chrovis/cljam/pull/209/checks?check_run_id=1060108149) 🤔

**Comment by athos** [comment]
Seems like a bug in clj-lint-action?
At a glance, [this line](https://github.com/xcoo/clj-lint-action/blob/1cc8b97397e053c1573531c37965aaec740069c0/lib/src/lint_action.clj#L111) should be `(cstr/join \: relative-files)` instead of `(cstr/join relative-files)`, I think.

**Comment by niyarin** [comment]
I'm sorry, I was going to check the newer clj-lint-action in the forked environment, but I made a mistake and pushed it here.
However, I found that the new clj-lint-action would fix bad annotations.
https://github.com/chrovis/cljam/runs/1061719609

I fixed the content of the lint-action comment.

**Comment by alumi** [comment]
@niyarin Thank you for updating `clj-lint-action`!

> I fixed the content of the lint-action comment.

But there're `are` and `testing` still referred but never used...?

**Comment by niyarin** [comment]
Sorry for the many mistakes :bow:
I fixed.

**Comment by alumi** [comment]
Never mind. Thank you! 👍

**Review by yito88 — APPROVED** [review]
LGTM👍

File: src/cljam/util/intervals.clj [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  [intervals]
+  (->> (group-by :chr intervals)
+       (into {}
+             (map
+              (fn [[chr xs]]
+                [chr
+                 (->> (group-by :start xs)
+                      (map (fn [[k vs]] [k (sort-by :end vs)]))
+                      (into (sorted-map)))])))
+       Sorted-map-intervals.))
```
Comment (alumi) [inline src/cljam/util/intervals.clj:26]: How about holding one record for each chromosome like the following?
This enables using different data structures based on a profile of intervals (e.g. no overlaps) in the future.

```clojure
(defprotocol IIntervals
  (find-overlap-intervals* [this start end]))

(deftype SortedMapIntervals [m]
  IIntervals
  (find-overlap-intervals* [this start end]
    (mapcat (fn [[_ x]] (drop-while #(< (:end %) start) x)) (subseq m <= end))))

(defn- make-sorted-map-intervals
  [intervals]
  (->> (group-by :start intervals)
       (map (fn [[k vs]] [k (sort-by :end vs)]))
       (into (sorted-map))
       (->SortedMapIntervals)))

(defn index-intervals
  [intervals]
  (->> intervals
       (group-by :chr)
       (into {} (map (fn [[k vs]] [k (make-sorted-map-intervals vs)])))))

(defn find-overlap-intervals
  [indexed-intervals chr start end]
  (find-overlap-intervals* (get indexed-intervals chr) start end))
```

File: src/cljam/util/intervals.clj [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+(defn index-intervals
+  "Make index for intervals to find overlaps."
+  [intervals]
+  (->> intervals
+       (group-by :chr)
+       (into {} (map (fn [[k vs]] [k (make-sorted-map-intervals vs)])))))
+
+(defn find-overlap-intervals
+  "Find intervals that is overlap query : chr and [start end]
+  from index data created by index-intervals."
```
Comment (alumi) [inline src/cljam/util/intervals.clj:29]:
```suggestion
  "Find intervals that are on the given `chr` and overlap the given interval
  [`start` `end`] using indexes created by `index-intervals`."
```

**Review by alumi — COMMENTED** [review]
Thank you! 👍 Let me comment a bit more on docstrings.

**Comment by niyarin** [comment]
Thank you for pointing.
I fixed them.

**Review by alumi — APPROVED** [review]
LGTM 👍

**Pull request merged**